**MAESTRÍA EN INTELIGENCIA ARTIFICIAL APLICADA**

**Curso: TC5053 - Ciencia y analítica de datos**

Tecnológico de Monterrey

Prof Grettel Barceló Alonso

**Semana 3**
Bases, almacenes y manipulación de datos

---

*   NOMBRE: Dario Martin Romero Cruz
*   MATRÍCULA: A01841015


---

En esta actividad usarás una base de datos relacional basada en el informe de participación y la lista del top 10 de Netflix. Incluye películas y programas de televisión, así como información sobre temporadas, métricas de visualización, fechas de estreno, duración y más, organizada en las siguientes tablas:

* `movie`: Información general de las películas.

* `tv_show`: Información general de los programas de televisión.

* `season`: Datos de las temporadas asociadas a cada programa de TV.

* `view_summary`: Métricas de visualización y rendimiento de películas o temporadas.

Revisa con detalle su esquema para que comprendas cómo se relacionan las tablas anteriores.

**NOTA IMPORTANTE:** Asegúrate de responder *explícitamente* todos los cuestionamientos.


`PyMySQL` es una librería escrita en Python puro que funciona como conector (*driver*) para motores de bases de datos MySQL, permitiendo abrir conexiones, ejecutar consultas SQL y recuperar resultados directamente desde programas en Python.

In [36]:
pip install pymysql

`SQLAlchemy` es una librería de Python que facilita la interacción con bases de datos y permite mantener un pool de conexiones eficiente, gestionar *commits* y *rollbacks* de forma automática y asegurar que múltiples conexiones simultáneas se manejen de manera segura, incluso cuando se ejecutan consultas SQL “puras”

In [37]:
# Importa las librerías necesarias
import pymysql
import sqlalchemy as sqla
import pandas as pd

Se crea una conexión (`conn`) para luego invocar declaraciones SQL.

In [38]:
# motor+driver://usuarioBD:clave@ipHostDBMS:puerto/esquema
# pool_recycle controla el tiempo máximo de vida de una conexión en el pool (3600 segundos = 1 hora)
db = sqla.create_engine('mysql+pymysql://mnaTC4029User:mnaTC4029Pass!@135.237.83.42:3306/Netflix', pool_recycle=3600)

Para que tus consultas sean más legibles y fáciles de mantener, puedes usar este formato multilínea con comillas triples y `sqla.text()`. Por ejemplo:

```
query = sqla.text("""
  SELECT ---
  FROM ---
  WHERE ---
""")
pd.read_sql(query, conn)
```

1.	Extrae toda la información de las películas que duran más de 5 horas.

In [39]:
#Definir consulta
query = sqla.text('SELECT * '
                  'FROM movie '
                  'WHERE runtime > 5 * 60')

#Conectar a BD
with db.connect() as conn:
    Data1 = pd.read_sql(query, conn)

#Imprimir resultados
print(f'Total de peliculas que duran mas de 5 hrs: {Data1.shape[0]}')
for i, Row in Data1.iterrows():
    print(f'  > {Row.title} - {int(Row.runtime/60)} horas {Row.runtime%60} minutos')

print('\nTodos los datos:\n')
print(Data1)

Total de peliculas que duran mas de 5 hrs: 5
  > Nihontouitsu Series: Film Series - 64 horas 52 minutos
  > Free and Easy Series: Film Series - 35 horas 20 minutos
  > Navarasa: Limited Series - 5 horas 12 minutos
  > Seiji Oda: Film Series - 11 horas 50 minutos
  > Kingdom ~ The Man Who Became the Top ~: Film Series - 7 horas 7 minutos

Todos los datos:

     id created_date modified_date available_globally locale  \
0  5793   2024-01-01    2024-01-01            b'\x00'   None   
1  5794   2024-01-01    2024-01-01            b'\x00'   None   
2  5874   2024-01-01    2024-01-01            b'\x01'   None   
3  9729   2024-01-01    2024-01-01            b'\x00'   None   
4  9730   2024-01-01    2024-01-01            b'\x00'   None   

           original_title release_date  runtime  \
0        日本統一シリーズ: 映画シリーズ         None     3892   
1          釣りバカ日誌: 映画シリーズ         None     2120   
2                    None   2021-08-06      312   
3      織田同志会 織田征仁: 映画シリーズ         None      710   
4 

2.	¿Cuál es el porcentaje de películas disponibles únicamente en EU en relación con el total, excluyendo los valores `NULL`? La consulta SQL debe entregar directamente el porcentaje final. No se deben devolver resultados parciales para realizar el cálculo en Python.

In [40]:
#Definir consulta
query = sqla.text('SELECT SUM(CASE WHEN available_globally = FALSE THEN 1 ELSE 0 END) / COUNT(*) * 100 '
                  'FROM movie '
                  'WHERE available_globally IS NOT NULL')

#Conectar a BD
with db.connect() as conn:
    Data2 = pd.read_sql(query, conn)

#Imprimir resultados
print(f'Peliculas disponibles solo en EU en relacion con el total: {Data2.iloc[0, 0]:.2f}%')

Peliculas disponibles solo en EU en relacion con el total: 83.57%


3.	¿Cuáles son los idiomas o regiones originales en la tabla de películas?
* ¿Cuántos registros tienen el campo `locale` con valor `NULL`? (NULL en SQL ⇔ None en Python)

In [41]:
#Definir consulta
query = sqla.text('SELECT locale '
                  'FROM movie')

#Conectar a BD
with db.connect() as conn:
    Data3 = pd.read_sql(query, conn)

#Imprimir resultados
print('Idiomas/regiones originales en la tabla de peliculas: ')
for v in Data3.loc[~Data3.locale.isna()].locale.unique():
    print('  > ' + v)

print(f'\nCantidad de registros con valor NULL en locale: {sum(Data3.locale.isna())}')

Idiomas/regiones originales en la tabla de peliculas: 
  > en

Cantidad de registros con valor NULL en locale: 11255


4.	Asumiendo que los valores `NULL` en `locale` corresponden a otro idioma (diferente del inglés), el título original de la película NO debería coincidir con el título principal en dichos registros.
* Determina cuántas películas tienen títulos diferentes en estos dos campos (`title` y `original_title`).
*  ¿Coinciden los resultados (cantidad de `NULL` y títulos diferentes)? Si no es así, identifica qué características tienen los registros restantes.
* Finalmente, concluye si la suposición de que los valores `NULL` en `locale` indican que la película está en otro idioma es válida.

In [50]:
query = sqla.text('SELECT * '
                  'FROM movie '
                  'WHERE title <> original_title '
                  'OR locale IS NULL')

#Conectar a BD
with db.connect() as conn:
    Data4 = pd.read_sql(query, conn)

#Imprimir resultados
print(f'Peliculas con diferente title y original_title: {sum(Data4.title != Data4.original_title)}')
print(f'Peliculas con valor NULL en locale: {sum(Data4.locale.isna())}')

print(f'\nPeliculas con valor NULL en title: {sum(Data4.title.isna())}')
print(f'Peliculas con valor NULL en original_title: {sum(Data4.original_title.isna())}')

print('\nAlgunas peliculas sin original_title:')
for v in list(Data4.loc[Data4.original_title.isna()].title[0:10]):
    print('  > ' + v)

Peliculas con diferente title y original_title: 11255
Peliculas con valor NULL en locale: 11255

Peliculas con valor NULL en title: 0
Peliculas con valor NULL en original_title: 7291

Algunas peliculas sin original_title:
  > Damsel
  > Lift
  > The Super Mario Bros. Movie
  > The Boss Baby
  > Leo
  > Trigger Warning
  > Sing 2
  > Heart of the Hunter
  > Despicable Me 2
  > Sing (2016)


CONCLUSION: Es cierto que si locale es NULL entonces los titulos no coinciden, esto sucede en 11255 registros. Sin embargo, no necesariamente la pelicula esta en otro idioma, porque la gran mayoria de estos casos (7291 registros) los titulos no coinciden dado que no se encuentra capturado el titulo original. Y viendo el titulo de las peliculas, podriamos suponer que si se encuentran en ingles. Es probable que cuando el titulo original sea NULL, entonces el valor de title corresponda al titulo original.

5.	Determina el título de la película que ha permanecido más tiempo en el top 10.

In [43]:
query = sqla.text('SELECT m.title, MAX(v.cumulative_weeks_in_top10) AS max_weeks '
                  'FROM movie m '
                  'JOIN view_summary v ON m.id = v.movie_id '
                  'GROUP BY m.id '
                  'ORDER BY max_weeks DESC '
                  'LIMIT 1')

#Conectar a BD
with db.connect() as conn:
    Data5 = pd.read_sql(query, conn)

#Imprimir resultados
print(f'Pelicula que ha permanecido mas tiempo en el top 10: {Data5.title[0]}')
print(f'Semanas: {Data5.max_weeks[0]}')

Pelicula que ha permanecido mas tiempo en el top 10: The Boss Baby
Semanas: 22


6.	Identifica los 5 programas de TV con mayor cantidad de temporadas.

In [44]:
query = sqla.text('SELECT t.title, COUNT(*) AS seasons '
                  'FROM tv_show t '
                  'JOIN season s ON t.id = s.tv_show_id '
                  'GROUP BY t.id '
                  'ORDER BY seasons DESC '
                  'LIMIT 5')

#Conectar a BD
with db.connect() as conn:
    Data6 = pd.read_sql(query, conn)

#Imprimir resultados
print('Programas de TV con mayor cantidad de temporadas disponibles:')
for i, Row in Data6.iterrows():
    print(f'  > {Row.title}: {Row.seasons} temporadas')

Programas de TV con mayor cantidad de temporadas disponibles:
  > Grey's Anatomy: 20 temporadas
  > Naruto Shippuden: 20 temporadas
  > Heartland (2007): 17 temporadas
  > It's Always Sunny in Philadelphia: 16 temporadas
  > Supernatural (2005): 15 temporadas


7.	¿Cuáles son los intervalos de fechas de los resúmenes en la tabla `view_summary` de los períodos (`duration`) semestrales?

In [45]:
query = sqla.text('SELECT DISTINCT start_date, end_date '
                  'FROM view_summary '
                  'WHERE duration = "SEMI_ANNUALLY" ')

#Conectar a BD
with db.connect() as conn:
    Data7 = pd.read_sql(query, conn)

#Imprimir resultados
print('Intervalos de fecha en los resumenes para periodos semestrales:')
for i, Row in Data7.iterrows():
    print(f'  > {Row.start_date:%d-%b-%y} a {Row.end_date:%d-%b-%y}')

Intervalos de fecha en los resumenes para periodos semestrales:
  > 01-ene.-24 a 30-jun.-24
  > 01-jul.-23 a 31-dic.-23


8.	Ordena las temporadas de *Grey's Anatomy* según la cantidad de vistas registradas en el primer período semestral de 2024.
* ¿Cómo interpretarías los resultados obtenidos?

In [46]:
query = sqla.text('SELECT s.season_number, v.views '
                  'FROM tv_show t '
                  'JOIN season s ON t.id = s.tv_show_id '
                  'JOIN view_summary v ON s.id = v.season_id '
                  'WHERE t.title = "Grey\'s Anatomy" '
                  'AND v.start_date = "2024-01-01" '
                  'AND v.end_date = "2024-06-30" '
                  'ORDER BY v.views DESC')

#Conectar a BD
with db.connect() as conn:
    Data8 = pd.read_sql(query, conn)

#Imprimir resultados
print('Temporadas de Grey\'s Anatomy con mayores vistas en el primer semestre de 2024:')
for i, Row in Data8.iterrows():
    print(f'  > Temp {Row.season_number}: {Row.views:,.0f} vistas')

Temporadas de Grey's Anatomy con mayores vistas en el primer semestre de 2024:
  > Temp 1: 3,600,000 vistas
  > Temp 2: 3,100,000 vistas
  > Temp 3: 2,900,000 vistas
  > Temp 5: 2,900,000 vistas
  > Temp 4: 2,900,000 vistas
  > Temp 6: 2,800,000 vistas
  > Temp 7: 2,700,000 vistas
  > Temp 8: 2,600,000 vistas
  > Temp 9: 2,500,000 vistas
  > Temp 10: 2,400,000 vistas
  > Temp 11: 2,200,000 vistas
  > Temp 12: 2,000,000 vistas
  > Temp 13: 2,000,000 vistas
  > Temp 14: 2,000,000 vistas
  > Temp 19: 2,000,000 vistas
  > Temp 15: 1,900,000 vistas
  > Temp 16: 1,800,000 vistas
  > Temp 18: 1,700,000 vistas
  > Temp 17: 1,700,000 vistas
  > Temp 20: 100,000 vistas


INTERPRETACIÓN: Se puede ver que las primeras temporadas tienen mayor cantidad de vistas. Sin embargo, esto puede deber a que son las que mas tiempo se han encontrado disponibles en la plataforma. Aun así hay que destacar algunas temporadas que no cumplen esta regla, como la temp 5 siendo más popular que la 4, o la 19 que logró superar las temporadas 15 a 18. Evidentemente la temporada 20 fue muy reciente, puesto que su número de vistas es demasiado bajo en comparación al resto.

9.	Todas las consultas anteriores podrían hacerse también con la lógica relacional implementada en Pandas, que permite replicar la mayoría de las operaciones de SQL. Si los dataframes se han llenado como sigue, resuelve la consulta 8 utilizando las funciones de Pandas.

In [47]:
conn = db.connect()
tv_show = pd.read_sql(sqla.text('SELECT * FROM tv_show'), conn)
season = pd.read_sql(sqla.text('SELECT * FROM season'), conn)
view_summary = pd.read_sql(sqla.text('SELECT * FROM view_summary'), conn)

In [48]:
import datetime

#Merge entre tabla tv_show (solo datos de Grey's Anatomy) y Season
Data9 = pd.merge(tv_show.loc[tv_show.title == 'Grey\'s Anatomy', ['id']].rename(columns = {'id': 'tv_show_id'}), season[['tv_show_id', 'id', 'season_number']], on = 'tv_show_id', how = 'inner')

#Merge entre tabla anterior y view_summary (solo datos del periodo seleccionado)
Data9 = pd.merge(Data9, view_summary.loc[(view_summary.start_date == datetime.date(2024, 1, 1)) & (view_summary.end_date == datetime.date(2024, 6, 30)), ['views', 'season_id']], left_on = 'id', right_on = 'season_id',  how = 'inner')

#Ordenar resultado
Data9.sort_values('views', ascending = False, ignore_index = True, inplace = True)

#Imprimir resultados
print('Temporadas de Grey\'s Anatomy con mayores vistas en el primer semestre de 2024:')
for i, Row in Data9.iterrows():
    print(f'  > Temp {Row.season_number}: {Row.views:,.0f} vistas')


Temporadas de Grey's Anatomy con mayores vistas en el primer semestre de 2024:
  > Temp 1.0: 3,600,000 vistas
  > Temp 2.0: 3,100,000 vistas
  > Temp 3.0: 2,900,000 vistas
  > Temp 5.0: 2,900,000 vistas
  > Temp 4.0: 2,900,000 vistas
  > Temp 6.0: 2,800,000 vistas
  > Temp 7.0: 2,700,000 vistas
  > Temp 8.0: 2,600,000 vistas
  > Temp 9.0: 2,500,000 vistas
  > Temp 10.0: 2,400,000 vistas
  > Temp 11.0: 2,200,000 vistas
  > Temp 12.0: 2,000,000 vistas
  > Temp 13.0: 2,000,000 vistas
  > Temp 14.0: 2,000,000 vistas
  > Temp 19.0: 2,000,000 vistas
  > Temp 15.0: 1,900,000 vistas
  > Temp 16.0: 1,800,000 vistas
  > Temp 18.0: 1,700,000 vistas
  > Temp 17.0: 1,700,000 vistas
  > Temp 20.0: 100,000 vistas


`MySQL` es un sistema de gestión de bases de datos relacional, pero desde Python también es posible extraer información de bases de datos no relacionales, como `Firestore`, `MongoDB` o `Cassandra`, utilizando conectores o integraciones específicas. Esto permite combinar datos de diferentes fuentes según las necesidades del análisis o la aplicación.

En el siguiente ejercicio usarás un ejemplo con `Firestore` desde Python. Para ello utilizarás los módulos `credentials` y `firestore` de la biblioteca `firebase_admin`.

In [52]:
#pip install firebase-admin

import pandas as pd
import firebase_admin
from firebase_admin import credentials
from firebase_admin import firestore

En `Firestore`, a diferencia de `MySQL` donde se utiliza un usuario y contraseña para conectarse, la autenticación se realiza mediante un archivo JSON que almacena las credenciales necesarias para acceder a la base de datos. Este archivo contiene las claves y la información de configuración que permiten a Python establecer la conexión de manera segura.

In [53]:
#from google.colab import drive
#drive.mount('/content/drive')

In [54]:
# Debes descargar el archivo de credenciales "consultancy.json" (disponible en las instrucciones de Canvas) y ubicarte en la carpeta donde lo almacenaste
import os
DIR = 'C:/Users/mromero01/Documents/GitHub/Tec_ENA/Ciencia y Analitica de Datos'
os.chdir(DIR)

#NOTA: La ejecucion de esta celda requiere acceso a una carpeta en mi equipo.

In [55]:
# consultancy.json almacena la clave privada para autenticar una cuenta y autorizar el acceso a los servicios
# A través de la función Certificate(), se regresa una credencial inicializada, que puedes utilizar para crear una nueva instancia de la aplicación
cred = credentials.Certificate('consultancy.json')
firebase_admin.initialize_app(cred)
db = firestore.client()

10.	Investiga cómo leer la colección `EMPLOYEE` y mostrar su contenido en un dataframe. Asegúrate de incluir el `id` en el resultado

In [56]:
#Obtener documentos de la colección
docs_stream = db.collection('EMPLOYEE').stream()

#Iterar para obtener la información
employees = []
for doc in docs_stream:
    data_doc = doc.to_dict()
    data_doc['id'] = doc.id #ID del registro
    employees.append(data_doc)

#Crear y formatear DataFrame final
df_employees = pd.DataFrame(employees).set_index('id')


#Imprimir resultados
print(f'Registros en la coleccion EMPLOYEE: {df_employees.shape[0]}\n')
print(df_employees)

Registros en la coleccion EMPLOYEE: 4

                       emp_lname              emp_hiredate emp_fname
id                                                                  
8LcLuxVHGAd2d9IQc5jF      Senior 1989-07-12 06:00:00+00:00     David
Fzd60D6Z2CU4n0wVV8YN  Smithfield 2004-06-04 05:00:00+00:00   William
lX5xuQ5w3i6ib2ExccWY        News 2000-11-08 06:00:00+00:00      John
yocFj2lichOkbAj9NBfp     Arbough 1996-12-01 06:00:00+00:00      June


In [ ]:
firebase_admin.delete_app(firebase_admin.get_app())

---

**Declaración de uso de IA**

Si aplica, deberá indicarse la herramienta y el modelo empleado en la entrega, así como la finalidad de su uso (generación de código / depuración / optimización).

Por ejemplo:

*   OpenAI. (2026). *ChatGPT (basado en GPT-4)* [Modelo de lenguaje grande], utilizado para generación de código y depuración. https://chat.openai.com/

---